src/lava/proc/dense/models.py

```diff
@@ -188,3 +188,148 @@ def run_spk(self):
+        This allows for the updating of the activation buffer and updating
+        weights.
+        """
+        return np.vstack([
+            np.where(self.delays == k, self.weights, 0)
+            for k in range(np.max(self.delays) + 1)
+        ])
+
+    def calc_act(self, s_in):
+        """
+        Calculate the activations by performing del_wgts * s_in. This matrix
+        is then summed across each row to get the activations to the output
+        neurons for different delays. This activation vector is reshaped to a
+        matrix of the form
+        (n_flat_output_neurons * (max_delay + 1), n_flat_output_neurons)
+        which is then transposed to get the activation matrix.
+        """
+        return np.reshape(np.sum(self.get_del_wgts() * s_in, axis=1),
+                          (np.max(self.delays) + 1, self.weights.shape[0])).T
+
+    def update_act(self, s_in):
+        """
+        Updates the activations for the connection.
+        Clears first column of a_buff and rolls them to the last column.
+        Finally, calculates the activations for the current time step and adds
+        them to a_buff.
+        This order of operations ensures that delays of 0 correspond to
+        the next time step.
+        """
+        self.a_buff[:, 0] = 0
+        self.a_buff = np.roll(self.a_buff, -1)
+        self.a_buff += self.calc_act(s_in)
+
+
+@implements(proc=DelayDense, protocol=LoihiProtocol)
+@requires(CPU)
+@tag("floating_pt")
+class PyDelayDenseModelFloat(AbstractPyDelayDenseModel):
+    """Implementation of Conn Process with Dense synaptic connections in
+    floating point precision. This short and simple ProcessModel can be used
+    for quick algorithmic prototyping, without engaging with the nuances of a
+    fixed point implementation. DelayDense incorporates delays into the Conn
+    Process.
+    """
+    s_in: PyInPort = LavaPyType(PyInPort.VEC_DENSE, bool, precision=1)
+    a_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, float)
+    a_buff: np.ndarray = LavaPyType(np.ndarray, float)
+    # weights is a 2D matrix of form (num_flat_output_neurons,
+    # num_flat_input_neurons) in C-order (row major).
+    weights: np.ndarray = LavaPyType(np.ndarray, float)
+    # delays is a 2D matrix of form (num_flat_output_neurons,
+    # num_flat_input_neurons) in C-order (row major).
+    delays: np.ndarray = LavaPyType(np.ndarray, int)
+    num_message_bits: np.ndarray = LavaPyType(np.ndarray, int, precision=5)
+
+    def run_spk(self):
+        # The a_out sent on a each timestep is a buffered value from dendritic
+        # accumulation at timestep t-1. This prevents deadlocking in
+        # networks with recurrent connectivity structures.
+        self.a_out.send(self.a_buff[:, 0])
+        if self.num_message_bits.item() > 0:
+            s_in = self.s_in.recv()
+        else:
+            s_in = self.s_in.recv().astype(bool)
+        self.update_act(s_in)
```

Comment:
Looks like `update_act` is called every time step, which in turn calls `calc_act`, which in turn calls `get_del_wgts`.
Couldn't we execute `get_del_wgts` just one time at initialization and store its result for later use ?
The computation happening there seem to be relatively heavy and might hinder performance if executed every time step.

Reply:
I actually tried that, but there are some arguments why this is not trivial to change. First, at initialization, the wgts and dlys are not numpy arrays yet, but LavaPyTypes - no operations work yet - this will change when the proc model is actually executed. More important though is the fact, that the wgts and maybe even the delays could change during execution. 

For now this seems to be the robust way to process delays.